classes/task/collect_metrics_task.php

```diff
@@ -0,0 +1,163 @@
+     * Display name for the task name.
+     *
+     * @return \lang_string|string
+     * @throws \coding_exception
+     */
+    public function get_name() {
+        return get_string('task_runner', 'tool_cloudmetrics');
+    }
+
+    /**
+     * Set a specific time for the use. Defaults to now.
+     *
+     * @param int $time
+     */
+    public function set_time(int $time) {
+        $this->time = $time;
+    }
+
+    /**
+     * Will send the metrics to the collectors.
+     *
+     * @param array $items
+     */
+    public function send_metrics(array $items) {
+        collector\manager::send_metrics($items);
+    }
+
+    /**
+     * Determine the highest level of frequency to be used.
+     *
+     * Returns the highest frequency for which a whole number of time intervals fits in the value given.
+     *
+     * Each interval (except monthly) is a whole number of the next smallest interval.
+     * So if (for example), it is determined that all 3 hour metrics are to be measured, all
+     * metrics for smaller intervals will also be measured.
+     *
+     * @param int $timediff The number of minutes since the reference time.
+     * @return int
+     */
+    public function get_frequency_cutoff(int $timediff): int {
+        if ($timediff % self::ONEWEEK == 0) {
+            return metric\manager::FREQ_WEEK;
+        } else if ($timediff % self::ONEDAY == 0) {
+            return metric\manager::FREQ_DAY;
+        } else if ($timediff % self::TWELVEHOURS == 0) {
+            return metric\manager::FREQ_12HOUR;
+        } else if ($timediff % self::THREEHOURS == 0) {
+            return metric\manager::FREQ_3HOUR;
+        } else if ($timediff % self::ONEHOUR == 0) {
+            return metric\manager::FREQ_HOUR;
+        } else if ($timediff % self::THIRTYMINUTES == 0) {
+            return metric\manager::FREQ_30MIN;
+        } else if ($timediff % self::FIFTEENMINUTES == 0) {
+            return metric\manager::FREQ_15MIN;
+        } else if ($timediff % self::FIVEMINUTES == 0) {
+            return metric\manager::FREQ_5MIN;
+        } else {
+            return metric\manager::FREQ_MIN;
+        }
+    }
+
+    /**
+     * Perform the task
+     *
+     * @throws \Exception
+     */
+    public function execute() {
+        // This algorithm is performance important. Any opportunity to optimize should be welcome.
+
+        // Use the server's timezone for determining times from strings.
+        $tz = \core_date::get_server_timezone_object();
+
+        // Get the time reference. The time string is in the server's timezone.
+        // It is then converted to a timestamp, which is UTC.
+        $reftime = (new \DateTime('midnight last Sunday', $tz))->getTimestamp();
+
+        // We don't need to use $tz here because timestamps are always UTC.
+        $time = $this->time ?? time();
+
+        // Get the number of minutes (rounded) that have passed since the ref time.
+        $timediff = (int) round(($time - $reftime) / 60);
```

Comment:
Can we replace all magic numbers here MINSECS or similar? makes it a little easier to read

Reply:
Done